build/generated.bzl

```diff
@@ -0,0 +1,149 @@
+def _format_output_dict(id, tags, output):
+    output.append("%s = {" % id)
+
+    for tag in tags.keys():
+        d = tags[tag]
+        output.append("%s%s: {" % (4 * " ", repr(tag)))
+        for k in d.keys():
+            output.append("%s%s: [" % (8 * " ", repr(k)))
+            for v in d[k].keys():
+                output.append("%s%s," % (12 * " ", repr(v)))
+            output.append("%s]," % (8 * " "))
+
+        output.append("%s}," % (4 * " "))
+
+    output.append("}\n")
+
+def _find_generator_tag_pkgs_impl(repo_ctx):
+    repo_root = repo_ctx.path(repo_ctx.attr._workspace).dirname
+
+    result = repo_ctx.execute(
+        ["grep", "--color=never", "-Hrox", "--include=*.go", "\\s*//\s*+k8s:\\S*=\\S*\\s*", repo_root],
+        quiet = True,
+    )
+    if result.return_code:
+        fail("failed searching for generator build tags: %s" % result.stderr)
+
+    # Maps tag names -> go packages -> values found for that tag in that package
+    tags_pkgs_values = {}
+
+    # Maps tag names -> values found for that tag -> go packages with those tag/value mappings
+    tags_values_pkgs = {}
+
+    for line in result.stdout.splitlines():
+        # Each line looks something like
+        # /a/full/path/to/file.go:+k8s:foo=bar
+
+        # First split out the filename from the matched tag blog
+        fname, _, match = line.partition(":")
+
+        # Remove the repo_root from the path, yielding just a relative path,
+        # and remove the basename to yield the package name.
+        pkg = fname[len("%s" % repo_root) + 1:].rpartition(sep = "/")[0]
+
+        # Skip things like _examples packages, since gazelle doesn't generate
+        # rules in these packages.
+        if pkg.startswith("_") or "/_" in pkg:
+            continue
+
+        # Strip off the leading "+k8s:", then split into tag name and value
+        tag, _, values = match.partition(":")[2].partition("=")
+
+        # The value may have multiple values, e.g. +k8s:foo=bar,baz, so handle each
+        # value separately
+        for value in values.split(","):
+            # Since Skylark doesn't have sets, we take the Go approach and fake a set using a dictionary
+            tags_pkgs_values.setdefault(tag, default = {}).setdefault(pkg, default = {})[value] = True
+            tags_values_pkgs.setdefault(tag, default = {}).setdefault(value, default = {})[pkg] = True
+
+    output = []
+    _format_output_dict("tags_pkgs_values", tags_pkgs_values, output)
+    _format_output_dict("tags_values_pkgs", tags_values_pkgs, output)
+
+    repo_ctx.file(
+        "tags.bzl",
+        content = "\n".join(output),
+        executable = False,
+    )
+
+    # Bazel needs a BUILD file even though nothing is in it
+    repo_ctx.file(
+        "BUILD.bazel",
+        content = "",
+        executable = False,
+    )
+
+    # Ensure that this rule always runs by touching the WORKSPACE file
+    repo_ctx.execute(["touch", repo_ctx.path(repo_ctx.attr._workspace)])
```

Comment:
I'm not sure a repository rule is the right initial approach here. Now every bazel command must first wait to recreate `tags.bzl`. How long does that take? It is not clear to me when bazel will start caching things -- aka is it smart enough to detect after recreating `tags.bzl` that the content is the same, and therefore used the cached `go_genrule()` output?

I think a better place to start is to make `hack/update-bazel.sh` update `tags.bzl`. And check in this `tags.bzl` file rather than generating it on the fly.

Reply:
I started by making kazel generate `tags.bzl` and checking this file into the repo, but I decided to re-implement as a repository rule for a few reasons:
* Having to run `hack/update-bazel.sh` is a source of developer friction. I thought our goal was to move away from extra out-of-band commands.
* Having a list of dependencies in one file is going to be a likely source of merge conflicts - not checking it in means it can't create conflicts.
* Our owners/approvals process is not well designed for having a single file that is updated frequently for distant changes. If we saved this as `build/tags.bzl`, then a variety of changes throughout the tree would require a `build/` owner to approve, adding more developer friction. 

A no-op `make bazel-build` seems to take about 6-7 seconds, exactly the same as current master, and doesn't actually rebuild everything. I think bazel may dump its analysis cache each time, but it's still smart enough to recognize that nothing needs to be rebuilt (unless the content of `tags.bzl` actually changes).

Reply:
I do wish there were a better way to tell bazel to run this rule every time, rather than touching `WORKSPACE` to force it to re-run. (I also considered having it touch this skylark file itself - `build/generated.bzl` - but the analysis time didn't seem significantly different, and it's still about as hacky.)

Reply:
this does seem pretty gross... ideally this could itself depend on the sources that invalidate it, but I guess that's maybe a lot anyhow..

Reply:
How else has agreed to maintain this? This is not something I am willing to maintain in its current form.

Can we at least move the logic to generate the tags.bzl file to outside to repository rule implementation?

I would like to see something like a standalone `generate_tags.py` script that accepts the necessary arguments and outputs this file. The repo impl can just be responsible for calling this script.

Once this script exists we can unit test it to ensure its functionality does not regress. And it is easier to move how/where/when the file is generated.

Reply:
I'm wary of adding outside dependencies like python, especially since k/k currently has no real python dependencies.

I'll refactor this and add some unit tests in skylark.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/hack/update_owners.py

Reply:
https://github.com/kubernetes/kubernetes/blob/master/hack/verify-flags-underscore.py

Reply:
given that this runs on every build invocation, I'd like to avoid invoking python if possible.